runtime/src/mailbox_api.rs

```diff
@@ -0,0 +1,147 @@
+// Licensed under the Apache-2.0 license
+
+use caliptra_drivers::{CaliptraError, CaliptraResult};
+use zerocopy::{AsBytes, FromBytes, LayoutVerified};
+
+#[derive(PartialEq, Eq)]
+pub struct CommandId(pub u32);
+
+impl CommandId {
+    pub const FIRMWARE_LOAD: Self = Self(0x46574C44); // "FWLD"
+    pub const GET_IDEV_CSR: Self = Self(0x49444556); // "IDEV"
+    pub const GET_LDEV_CERT: Self = Self(0x4C444556); // "LDEV"
+    pub const ECDSA384_VERIFY: Self = Self(0x53494756); // "SIGV"
+    pub const STASH_MEASUREMENT: Self = Self(0x4D454153); // "MEAS"
+    pub const INVOKE_DPE: Self = Self(0x44504543); // "DPEC"
+
+    pub const TEST_ONLY_GET_LDEV_CERT: Self = Self(0x4345524c); // "CERL"
+    pub const TEST_ONLY_GET_FMC_ALIAS_CERT: Self = Self(0x43455246); // "CERF"
+
+    /// FIPS module commands.
+    /// The status command.
+    pub const VERSION: Self = Self(0x4650_5652); // "FPVR"
+    /// The self-test command.
+    pub const SELF_TEST: Self = Self(0x4650_4C54); // "FPST"
+    /// The shutdown command.
+    pub const SHUTDOWN: Self = Self(0x4650_5344); // "FPSD"
+}
+impl From<u32> for CommandId {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+impl From<CommandId> for u32 {
+    fn from(value: CommandId) -> Self {
+        value.0
+    }
+}
+
+// Helpers
+// Assumes given byte array is >= size_of T
+// Will return CaliptraError::RUNTIME_INTERNAL otherwise
+pub fn cast_bytes_to_struct<T: AsBytes + FromBytes>(bytes: &[u8]) -> CaliptraResult<&T> {
+    Ok(
+        LayoutVerified::<&[u8], T>::new(&bytes[..core::mem::size_of::<T>()])
```

Comment:
can't just use zerocopy api?

Reply:
This is a good question, I am hoping there is a better way to do this. So LayoutVerified is part of the zerocopy crate. But most of what I think you would call the actual API for the crate seems to only support creating _copies_ from struct to buffer or vice-versa instead of working on the same buffer in memory.

So this LayoutVerified::new was the only way I found to handle that. Then it was a bit messy so I just wrapped it in this function.

Reply:
One small correction on my previous comment: It is supported to get a reference to a byte array _from_ a struct using as_bytes() or as_bytes_mut(). It's just the other direction I could not find a cleaner solution for in zerocopy or elsewhere.

Reply:
No problem, I can get rid of this